File: backend/src/main/java/ru/otus/backend/db/entity/Customer.java
package ru.otus.backend.db.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Transient;
import org.springframework.data.relational.core.mapping.Table;

import static java.util.Optional.ofNullable;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Table("customers")
public class Customer {

    @Id
    private Long id;
    private String name;
    private String address;
    private Long userId;
    private Long billingAccountId;
    private Status status;

    @Transient
    private User user;
    @Transient
    private BillingAccount billingAccount;

    public Customer(Long id, String name, String address) {
        this.id = id;
        this.name = name;
        this.address = address;
    }

    public Long getUserId() {
        return ofNullable(userId).or(() -> ofNullable(user).map(User::getId)).orElse(null);
    }

    public void setUser(User user) {
        this.userId = ofNullable(user).map(User::getId).orElse(null);
        this.user = user;
    }

    public void setBillingAccount(BillingAccount billingAccount) {
        this.billingAccountId = ofNullable(billingAccount).map(BillingAccount::getId).orElse(null);
        this.billingAccount = billingAccount;
    }
}

File: frontend/src/app/main-module/models/customer.ts
import {User} from "./user";
import {BillingAccount} from "./billing-account";
import {Status} from "./status";

export class Customer {
    id: string;
    name: string;
    address: string;
    user: User = new User();
    billingAccount: BillingAccount;
    status: Status = new Status();

    constructor() {
        if (localStorage.getItem('wallet')) {
            this.billingAccount = new BillingAccount();
        }
    }

    static cloneCustomer(customer: Customer): Customer {
        let cloneCustomer: Customer = new Customer();
        cloneCustomer.id = customer.id;
        cloneCustomer.name = customer.name;
        cloneCustomer.address = customer.address;
        cloneCustomer.user = customer.user;
        cloneCustomer.billingAccount = customer.billingAccount;
        cloneCustomer.user = User.cloneUser(customer.user);
        cloneCustomer.status = Status.cloneStatus(customer.status);
        return cloneCustomer;
    }
}
